File: solutions/python/pig-latin/3/pig_latin.py
from typing import Final

def _space_word(substr: str) -> tuple[str, str]:
    sep: Final[int] = next((
        index for index in range(len(substr)) if not substr[index].isspace()
    ), len(substr))
    return substr[:sep], substr[sep:]

def _word_space(substr: str) -> tuple[str, str]:
    sep: Final[int] = next((
        index for index in range(len(substr)) if substr[index].isspace()
    ), len(substr))
    return substr[:sep], substr[sep:]
    
def _cons_vowel(substr: str) -> tuple[str, str]:
    sep: Final[int] = next((
        index for index in range(len(substr)) if substr[index].lower() in 'aeiou'
    ), len(substr))
    return substr[:sep], substr[sep:]

def translate(text: Final[str]) -> str:
    subtext: str = text[:]
    result: list[str] = []
    while len(subtext) > 0:
        if subtext[0].isspace():
            spaces, subtext = _space_word(subtext)
            result.append(spaces)
            continue
        substr, subtext = _word_space(subtext)
        cons, vowel = _cons_vowel(substr)
        if len(cons) == 0 or cons.startswith('xr') or cons.startswith('yt'):
            result.extend((substr, 'ay'))
            continue
        ypos: Final[int] = cons.find('y')
        if ypos > 0:
            result.extend((substr[ypos:], substr[:ypos], 'ay'))
            continue
        if cons.endswith('q') and vowel.startswith('u'):
            result.extend((vowel[1:], substr[:len(cons) + 1], 'ay'))
            continue
        result.extend((vowel, cons, 'ay'))
    return ''.join(result)

File: solutions/go/pig-latin/3/pig_latin.go
package piglatin

import (
    "strings"
    "unicode"
    "unicode/utf8"
)

func Sentence(sentence string) string {
	var b strings.Builder
    for len(sentence) > 0 {
        first, _ := utf8.DecodeRuneInString(sentence)
        if unicode.IsSpace(first) {
            var spaces string
            spaces, sentence = spaceWord(sentence)
            b.WriteString(spaces)
            continue
        }
        var substr string
        substr, sentence = wordSpace(sentence)
        cons, vowel := consVowel(substr)
        if len(cons) == 0 || strings.HasPrefix(cons, "xr") || strings.HasPrefix(cons, "yt") {
            b.WriteString(substr)
            b.WriteString("ay")
            continue
        }
        ypos := strings.Index(cons, "y")
        if ypos > 0 {
            b.WriteString(substr[ypos:])
            b.WriteString(substr[:ypos])
            b.WriteString("ay")
            continue
        }
        if strings.HasSuffix(cons, "q") && strings.HasPrefix(vowel, "u") {
            b.WriteString(vowel[1:])
            b.WriteString(cons)
            b.WriteString(vowel[:1])
            b.WriteString("ay")
            continue
        }
        b.WriteString(vowel)
        b.WriteString(cons)
        b.WriteString("ay")
    }
    return b.String()
}

func sepUtil(substr string, sepAt func(rune) bool) (string, string) {
    sep := strings.IndexFunc(substr, sepAt)
    if sep == -1 {
        sep = len(substr)
    }
    return substr[:sep], substr[sep:]
}

func spaceWord(substr string) (string, string) {
    return sepUtil(substr, isWord)
}

func isWord(r rune) bool {
    return !unicode.IsSpace(r)
}

func wordSpace(substr string) (string, string) {
    return sepUtil(substr, unicode.IsSpace)
}

func consVowel(substr string) (string, string) {
    return sepUtil(substr, isVowel)
}

func isVowel(r rune) bool {
	switch r {
	case 'a', 'e', 'i', 'o', 'u':
		return true
    default:
        return false
	}
}